File: src/fr/cph/chicago/app/activity/TrainMapActivity.java
/**
 * Copyright 2016 Carl-Philipp Harmant
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.cph.chicago.app.activity;

import android.app.Activity;
import android.app.FragmentManager;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.annimon.stream.Stream;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.GoogleMapOptions;
import com.google.android.gms.maps.MapFragment;
import com.google.android.gms.maps.MapsInitializer;
import com.google.android.gms.maps.model.BitmapDescriptor;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.android.gms.maps.model.PolylineOptions;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import fr.cph.chicago.app.App;
import fr.cph.chicago.R;
import fr.cph.chicago.data.DataHolder;
import fr.cph.chicago.data.TrainData;
import fr.cph.chicago.entity.Position;
import fr.cph.chicago.entity.Train;
import fr.cph.chicago.entity.enumeration.TrainLine;
import fr.cph.chicago.app.listener.TrainMapOnCameraChangeListener;
import fr.cph.chicago.app.task.LoadTrainFollowTask;
import fr.cph.chicago.app.task.LoadTrainPositionTask;
import fr.cph.chicago.util.Util;

/**
 * @author Carl-Philipp Harmant
 * @version 1
 */
public class TrainMapActivity extends Activity {

    private ViewGroup viewGroup;
    private MapFragment mapFragment;
    private Marker selectedMarker;
    private Map<Marker, View> views;
    private RelativeLayout layout;
    private String line;
    private Map<Marker, Boolean> status;
    private List<Marker> markers;
    private TrainData trainData;
    private TrainMapOnCameraChangeListener trainListener;

    private boolean centerMap = true;
    private boolean refreshingInfoWindow = false;
    private boolean drawLine = true;

    public TrainMapActivity() {
        this.views = new HashMap<>();
    }

    @Override
    public final void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        App.checkTrainData(this);
        if (!this.isFinishing()) {
            MapsInitializer.initialize(getApplicationContext());
            setContentView(R.layout.activity_map);
            layout = (RelativeLayout) findViewById(R.id.map);
            viewGroup = (ViewGroup) findViewById(android.R.id.content);
            if (savedInstanceState != null) {
                line = savedInstanceState.getString(getString(R.string.bundle_train_line));
            } else {
                line = getIntent().getExtras().getString(getString(R.string.bundle_train_line));
            }

            // Init data
            initData();

            // Init toolbar
            setToolbar();

            // Google analytics
            Util.trackScreen(getApplicationContext(), getString(R.string.analytics_train_map));
        }
    }

    private void initData() {
        // Load data
        final DataHolder dataHolder = DataHolder.getInstance();
        trainData = dataHolder.getTrainData();
        markers = new ArrayList<>();
        status = new HashMap<>();
        trainListener = new TrainMapOnCameraChangeListener(getApplicationContext());
    }

    private void setToolbar() {
        final Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        toolbar.inflateMenu(R.menu.main);
        toolbar.setOnMenuItemClickListener((item -> {
            new LoadTrainPositionTask(TrainMapActivity.this, line, trainData).execute(false, true);
            return false;
        }));
        final TrainLine trainLine = TrainLine.fromXmlString(line);
        Util.setWindowsColor(this, toolbar, trainLine);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            toolbar.setElevation(4);
        }
        toolbar.setNavigationIcon(R.drawable.ic_arrow_back_white_24dp);
        toolbar.setOnClickListener(v -> finish());

        toolbar.setTitle(trainLine.toStringWithLine());
    }

    @Override
    public final void onStart() {
        super.onStart();
        if (mapFragment == null) {
            final FragmentManager fm = getFragmentManager();
            mapFragment = (MapFragment) fm.findFragmentById(R.id.map);
            final GoogleMapOptions options = new GoogleMapOptions();
            final CameraPosition camera = new CameraPosition(Util.CHICAGO, 10, 0, 0);
            options.camera(camera);
            mapFragment = MapFragment.newInstance(options);
            mapFragment.setRetainInstance(true);
            fm.beginTransaction().replace(R.id.map, mapFragment).commit();
        }
    }

    @Override
    public final void onStop() {
        super.onStop();
        centerMap = false;
    }

    @Override
    public final void onResume() {
        super.onResume();
        mapFragment.getMapAsync(googleMap -> {
            Util.setLocationOnMap(this, googleMap);
            googleMap.setInfoWindowAdapter(new GoogleMap.InfoWindowAdapter() {
                @Override
                public View getInfoWindow(final Marker marker) {
                    return null;
                }

                @Override
                public View getInfoContents(final Marker marker) {
                    if (!"".equals(marker.getSnippet())) {
                        final View view = views.get(marker);
                        if (!refreshingInfoWindow) {
                            selectedMarker = marker;
                            final String runNumber = marker.getSnippet();
                            new LoadTrainFollowTask(TrainMapActivity.this, view, false, trainData).execute(runNumber);
                            status.put(marker, false);
                        }
                        return view;
                    } else {
                        return null;
                    }
                }
            });

            googleMap.setOnInfoWindowClickListener(new GoogleMap.OnInfoWindowClickListener() {
                @Override
                public void onInfoWindowClick(final Marker marker) {
                    if (!"".equals(marker.getSnippet())) {
                        final View view = views.get(marker);
                        if (!refreshingInfoWindow) {
                            selectedMarker = marker;
                            final String runNumber = marker.getSnippet();
                            final Boolean current = status.get(marker);
                            new LoadTrainFollowTask(TrainMapActivity.this, view, !current, trainData).execute(runNumber);
                            status.put(marker, !current);
                        }
                    }
                }
            });
            if (Util.isNetworkAvailable(getApplicationContext())) {
                new LoadTrainPositionTask(TrainMapActivity.this, line, trainData).execute(centerMap, true);
            } else {
                Util.showNetworkErrorMessage(layout);
            }
        });
    }

    @Override
    public void onRestoreInstanceState(final Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        line = savedInstanceState.getString(getString(R.string.bundle_train_line));
    }

    @Override
    public void onSaveInstanceState(final Bundle savedInstanceState) {
        savedInstanceState.putString(getString(R.string.bundle_train_line), line);
        super.onSaveInstanceState(savedInstanceState);
    }

    public void refreshInfoWindow() {
        if (selectedMarker == null) {
            return;
        }
        refreshingInfoWindow = true;
        selectedMarker.showInfoWindow();
        refreshingInfoWindow = false;
    }

    public void centerMapOnTrain(@NonNull final List<Train> result) {
        mapFragment.getMapAsync(googleMap -> {
            final Position position;
            final int zoom;
            if (result.size() == 1) {
                position = result.get(0).getPosition();
                zoom = 15;
            } else {
                position = Train.getBestPosition(result);
                zoom = 11;
            }
            final LatLng latLng = new LatLng(position.getLatitude(), position.getLongitude());
            googleMap.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng, zoom));
        });
    }

    public void drawTrains(@NonNull final List<Train> trains) {
        mapFragment.getMapAsync(googleMap -> {
            if (views == null) {
                views = new HashMap<>();
            } else {
                views.clear();
            }
            for (final Marker marker : markers) {
                marker.remove();
            }
            markers.clear();
            final BitmapDescriptor bitmapDescr = trainListener.getCurrentBitmapDescriptor();
            for (final Train train : trains) {
                final LatLng point = new LatLng(train.getPosition().getLatitude(), train.getPosition().getLongitude());
                final String title = "To " + train.getDestName();
                final String snippet = Integer.toString(train.getRouteNumber());

                final Marker marker = googleMap.addMarker(new MarkerOptions().position(point).title(title).snippet(snippet).icon(bitmapDescr).anchor(0.5f, 0.5f).rotation(train.getHeading()).flat(true));
                markers.add(marker);

                final View view = TrainMapActivity.this.getLayoutInflater().inflate(R.layout.marker_train, viewGroup, false);
                final TextView title2 = (TextView) view.findViewById(R.id.title);
                title2.setText(title);

                final TextView color = (TextView) view.findViewById(R.id.route_color_value);
                color.setBackgroundColor(TrainLine.fromXmlString(line).getColor());

                views.put(marker, view);
            }

            trainListener.setTrainMarkers(markers);

            googleMap.setOnCameraChangeListener(trainListener);
        });
    }

    public void drawLine(@NonNull final List<Position> positions) {
        if (drawLine) {
            mapFragment.getMapAsync(googleMap -> {
                final PolylineOptions poly = new PolylineOptions();
                poly.width(7f);
                poly.geodesic(true).color(TrainLine.fromXmlString(line).getColor());
                Stream.of(positions)
                    .map(position -> new LatLng(position.getLatitude(), position.getLongitude()))
                    .forEach(poly::add);

                googleMap.addPolyline(poly);
            });
            drawLine = false;
        }
    }
}

File: android-app/src/googleplay/kotlin/fr/cph/chicago/core/activity/map/TrainMapActivity.kt
/**
 * Copyright 2019 Carl-Philipp Harmant
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.cph.chicago.core.activity.map

import android.annotation.SuppressLint
import android.graphics.BitmapFactory
import android.os.Bundle
import android.view.View
import android.widget.TextView
import butterknife.BindString
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.GoogleMap.InfoWindowAdapter
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.Marker
import com.google.android.gms.maps.model.MarkerOptions
import com.google.android.gms.maps.model.PolylineOptions
import fr.cph.chicago.R
import fr.cph.chicago.core.App
import fr.cph.chicago.core.model.Position
import fr.cph.chicago.core.model.Train
import fr.cph.chicago.core.model.enumeration.TrainLine
import fr.cph.chicago.core.model.marker.RefreshTrainMarkers
import fr.cph.chicago.rx.TrainEtaObserver
import fr.cph.chicago.service.TrainService
import fr.cph.chicago.util.MapUtil
import fr.cph.chicago.util.Util
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.rxkotlin.Singles
import io.reactivex.schedulers.Schedulers
import org.apache.commons.lang3.StringUtils
import timber.log.Timber

/**
 * @author Carl-Philipp Harmant
 * @version 1
 */
class TrainMapActivity : FragmentMapActivity() {

    companion object {
        private val util = Util
        private val trainService = TrainService
    }

    @BindString(R.string.bundle_train_line)
    lateinit var bundleTrainLine: String

    private var views: MutableMap<Marker, View> = hashMapOf()

    private lateinit var line: String
    private lateinit var refreshTrainMarkers: RefreshTrainMarkers
    private var status: MutableMap<Marker, Boolean> = mutableMapOf()
    private var trainsMarker: List<Marker> = listOf()
    private val stationMarkers: MutableList<Marker> = mutableListOf()

    private var centerMap = true
    private var drawLine = true

    override fun create(savedInstanceState: Bundle?) {
        line = if (savedInstanceState != null)
            savedInstanceState.getString(bundleTrainLine) ?: StringUtils.EMPTY
        else
            intent.getStringExtra(bundleTrainLine) ?: StringUtils.EMPTY

        // Init data
        initData()

        // Init toolbar
        setToolbar()
    }

    override fun setToolbar() {
        super.setToolbar()
        toolbar.setOnMenuItemClickListener {
            centerMap = false
            loadActivityData()
            false
        }

        val trainLine = TrainLine.fromXmlString(line)
        util.setWindowsColor(this, toolbar, trainLine)
        toolbar.title = trainLine.toStringWithLine()
    }

    public override fun onStop() {
        super.onStop()
        centerMap = false
    }

    public override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        line = savedInstanceState.getString(bundleTrainLine) ?: StringUtils.EMPTY
    }

    public override fun onSaveInstanceState(savedInstanceState: Bundle) {
        if (::line.isInitialized) savedInstanceState.putString(bundleTrainLine, line)
        super.onSaveInstanceState(savedInstanceState)
    }

    private fun centerMapOnTrain(result: List<Train>) {
        val position: Position
        val zoom: Int
        if (result.size == 1) {
            position = if (result[0].position.latitude == 0.0 && result[0].position.longitude == 0.0)
                MapUtil.chicagoPosition
            else
                result[0].position
            zoom = 15
        } else {
            position = MapUtil.getBestPosition(result.map { it.position })
            zoom = 11
        }
        centerMapOn(position.latitude, position.longitude, zoom)
    }

    private fun drawTrains(trains: List<Train>) {
        resetData()
        val bitmapDesc = refreshTrainMarkers.currentDescriptor
        trainsMarker = trains.map { (routeNumber, destName, _, position, heading) ->
            val point = LatLng(position.latitude, position.longitude)
            val title = "To $destName"
            val snippet = routeNumber.toString()
            googleMap.addMarker(MarkerOptions().position(point).title(title).snippet(snippet).icon(bitmapDesc).anchor(0.5f, 0.5f).rotation(heading.toFloat()).flat(true))
        }.onEach { marker ->
            val view = layoutInflater.inflate(R.layout.marker, viewGroup, false)
            val title2 = view.findViewById<TextView>(R.id.title)
            title2.text = marker.title
            views[marker] = view
        }
    }

    private fun resetData() {
        views.clear()
        trainsMarker.forEach { it.remove() }
    }

    private fun drawLine(positions: List<Position>) {
        val poly = PolylineOptions()
        poly.width((application as App).lineWidthGoogleMap)
        poly.geodesic(true).color(TrainLine.fromXmlString(line).color)
        positions
            .map { position -> LatLng(position.latitude, position.longitude) }
            .forEach { poly.add(it) }

        googleMap.addPolyline(poly)
        drawLine = false
    }

    override fun onCameraIdle() {
        refreshTrainMarkers.refreshTrainAndStation(googleMap.cameraPosition, trainsMarker, stationMarkers)
    }

    override fun onMapReady(googleMap: GoogleMap) {
        super.onMapReady(googleMap)

        refreshTrainMarkers = RefreshTrainMarkers()

        Observable.fromCallable { BitmapDescriptorFactory.fromBitmap(BitmapFactory.decodeResource(resources, colorDrawable())) }
            .flatMapIterable { icon -> trainService.getStationsForLine(TrainLine.fromXmlString(line)).map { Pair(it, icon) } }
            .map { pair ->
                val station = pair.first
                val point = LatLng(station.stopsPosition[0].latitude, station.stopsPosition[0].longitude)
                MarkerOptions()
                    .position(point)
                    .title(station.name)
                    .icon(pair.second)
            }
            .subscribeOn(Schedulers.computation())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                { markerOptions ->
                    val marker = googleMap.addMarker(markerOptions)
                    marker.isVisible = false
                    stationMarkers.add(marker)
                },
                { error -> Timber.e(error) })

        googleMap.setInfoWindowAdapter(object : InfoWindowAdapter {
            override fun getInfoWindow(marker: Marker): View? {
                return null
            }

            override fun getInfoContents(marker: Marker): View? {
                return when {
                    stationMarkers.contains(marker) -> null
                    StringUtils.EMPTY != marker.snippet -> {
                        val view = views[marker]
                        if (!refreshingInfoWindow) {
                            selectedMarker = marker
                            val runNumber = marker.snippet
                            trainService.trainEtas(runNumber, false)
                                .subscribe(TrainEtaObserver(view!!, this@TrainMapActivity))
                            status[marker] = false
                        }
                        view
                    }
                    else -> null
                }
            }
        })
        googleMap.setOnInfoWindowClickListener { marker ->
            if (StringUtils.EMPTY != marker.snippet) {
                val view = views[marker]
                if (!refreshingInfoWindow) {
                    selectedMarker = marker
                    val runNumber = marker.snippet
                    val current = status[marker] ?: false
                    trainService.trainEtas(runNumber, !current)
                        .subscribe(TrainEtaObserver(view!!, this@TrainMapActivity))
                    status[marker] = !current
                }
            }
        }
        loadActivityData()
    }

    @SuppressLint("CheckResult")
    private fun loadActivityData() {
        // Load train location
        val trainsSingle = trainService.trainLocations(line)
        // Load pattern from local file
        val positionsSingle = trainService.readPatterns(TrainLine.fromXmlString(line))

        if (drawLine) {
            Singles.zip(trainsSingle, positionsSingle, zipper = { trains, positions ->
                drawTrains(trains)
                drawLine(positions.map { it.position })
                if (trains.isNotEmpty()) {
                    if (centerMap) {
                        centerMapOnTrain(trains)
                    }
                } else {
                    util.showSnackBar(this@TrainMapActivity.currentFocus!!, R.string.message_no_train_found)
                }
                Any()
            }).subscribe(
                {},
                { throwable ->
                    util.showSnackBar(this@TrainMapActivity.currentFocus!!, R.string.message_error_while_loading_data)
                    Timber.e(throwable, "Error not handled")
                }
            )
        } else {
            trainsSingle.subscribe(
                { trains ->
                    if (trains != null) {
                        drawTrains(trains)
                        if (trains.isEmpty()) {
                            util.showSnackBar(this@TrainMapActivity.currentFocus!!, R.string.message_no_train_found)
                        }
                    } else {
                        util.showSnackBar(this@TrainMapActivity.currentFocus!!, R.string.message_error_while_loading_data)
                    }
                },
                { error -> Timber.e(error) })
        }
    }

    private fun colorDrawable(): Int {
        return when (TrainLine.fromXmlString(line)) {
            TrainLine.BLUE -> R.drawable.blue_marker_no_shade
            TrainLine.BROWN -> R.drawable.brown_marker_no_shade
            TrainLine.GREEN -> R.drawable.green_marker_no_shade
            TrainLine.ORANGE -> R.drawable.orange_marker_no_shade
            TrainLine.PINK -> R.drawable.pink_marker_no_shade
            TrainLine.PURPLE -> R.drawable.purple_marker_no_shade
            TrainLine.RED -> R.drawable.red_marker_no_shade
            TrainLine.YELLOW -> R.drawable.yellow_marker
            TrainLine.NA -> R.drawable.red_marker_no_shade
        }
    }
}
